diff --git a/Lib/LibStatic/CMakeLists.txt b/Lib/LibStatic/CMakeLists.txt
index c116ee5..1a4d14c 100644
--- a/Lib/LibStatic/CMakeLists.txt
+++ b/Lib/LibStatic/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 2.6)
+cmake_minimum_required(VERSION 3.5)
 project(RakNetLibStatic)

 FILE(GLOB ALL_HEADER_SRCS ${RakNet_SOURCE_DIR}/Source/*.h)
@@ -12,6 +12,7 @@ IF(WIN32 AND NOT UNIX)
 	SET( CMAKE_CXX_FLAGS "/D WIN32 /D _RAKNET_LIB /D _CRT_NONSTDC_NO_DEPRECATE /D _CRT_SECURE_NO_DEPRECATE /GS- /GR- ")
 ENDIF(WIN32 AND NOT UNIX)

+IF(0)
 IF(WIN32 AND NOT UNIX)
 	target_link_libraries (RakNetLibStatic ${RAKNET_LIBRARY_LIBS})

@@ -30,5 +31,13 @@ ELSE(WIN32 AND NOT UNIX)
 	INSTALL(TARGETS RakNetLibStatic DESTINATION ${RakNet_SOURCE_DIR}/Lib/RakNetLibStatic)
 	INSTALL(FILES ${ALL_HEADER_SRCS} DESTINATION ${RakNet_SOURCE_DIR}/include/raknet)
 ENDIF(WIN32 AND NOT UNIX)
-
-
+ENDIF()
+
+target_link_libraries(RakNetLibStatic ${RAKNET_LIBRARY_LIBS})
+set_target_properties(RakNetLibStatic PROPERTIES OUTPUT_NAME raknet-static)
+include(GNUInstallDirs)
+install(TARGETS RakNetLibStatic
+        ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
+install(FILES ${ALL_HEADER_SRCS} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/raknet)
